### ISSUE TITLE
2-factor verification for administration tools

### ISSUE THREAD

**Issue by Ellyster** [rethinkdb/rethinkdb#3589]
It would be great if RethinkDB allowed db admins to configure a 2-factor verification login for the administration tools (both web interface and cli), in that way it would be safer to run a RethinkDB cluster on an open network.

Some candidates:
- [Authy](https://www.authy.com/)
- [Google Authenticator](https://github.com/google/google-authenticator)

**Comment by kofalt** [comment]
From the [security documentation](http://www.rethinkdb.com/docs/security/), it seems they largely rely on correctly configured networks:

> The best way to secure a RethinkDB cluster is to run it on a protected network that doesn't allow access from the outside world. 

There is an optional authentication system, but (probably deliberately) kept very simple:

> Note: the authentication key will be transmitted to and stored on the RethinkDB server in plain text, and neither the key nor the data passed between the client and the server will be encrypted. The key provides basic protection against unauthorized access, but if the client port is open to outside networks it's strongly suggested you use SSH tunneling for protection (see below).

IMO this is absolutely the right way to do it. The network is a powerful access management tool available to an administrator, and should be treated accordingly. The alternative is only selecting systems that all attempt to implement something that is not their core competency.

Other examples of excellent software taking the same approach include the [redis security docs](http://redis.io/topics/security):

> Redis is designed to be accessed by trusted clients inside trusted environments. This means that usually it is not a good idea to expose the Redis instance directly to the internet or, in general, to an environment where untrusted clients can directly access the Redis TCP port or UNIX socket.

This matches how I expect competent systems to be designed: keep your [attack surface small](http://www.sans.edu/research/security-laboratory/article/did-attack-surface), spend a lot of effort securing your attack surface, and everything else goes behind access restrictions. This is defence in depth 101, and is a battle-tested & industry-standard security doctrine.

In summary I would disagree with development effort being spent on a feature like this because it runs counter to my understanding of best practices for secure environments. That all said, I'm not employed by RethinkDB, and have no say in the matter :)

**Comment by danielmewes** [comment]
Thank you for the suggestion @Ellyster .
We're planning to make setting up SSL/TLS and/or other tunnels much easier in the future.
See https://github.com/rethinkdb/rethinkdb/issues/3151 and https://github.com/rethinkdb/rethinkdb/issues/3158

I'm not sure yet whether this is going to include the web administration UI, but we'll consider it. The suggested approaches in the other threads would be based on a private/public key authentication scheme though, not 2-factor authentication. Would that work for you?

**Comment by Ellyster** [comment]
Maybe it is easier if I explain our situation.

We are working in an application to build collaborative repository of open access research literature, and thus we need to allow each university to deploy a RethinkDB node in their own servers and join to the cluster network.

So they will have direct access to the each hosting server (it makes network-based security useless), but they shouldn't be allowed to access the db management tools (even if they have local access to the server).

So as long as there is a way to disallow a new node to join the cluster if it hasn't the right security configuration (only our admin keys) for the management panel a Private/public key authentication scheme will work for me too.

**Comment by danielmewes** [comment]
@Ellyster thank you for your explanation.

Please keep in mind that once a server has joined a cluster, it always technically has full access to the cluster configuration. In addition, every user can delete tables and databases through ReQL, even without access to the administration tools.
Please also note that all administration tasks will be performable directly through ReQL queries starting with the upcoming release (1.16). So everyone who runs queries can also change the cluster configuration, unless you implement additional filtering through some sort of ReQL proxy on your side.

**Comment by frankleng** [comment]
+1

**Comment by danesparza** [comment]
+1
